CIP2/CIP2.md

```diff
@@ -0,0 +1,1001 @@
+is more than enough to cover a given payment amount, if we attempt to include
+that dust in a given transaction, we may run out of space (by reaching the
+[transaction size limit](#the-transaction-size-limitation)) before we can cover
+the target amount.
+
+# Interface
+
+All coin selection algorithms used by Cardano Wallet implement a
+_common interface_.
+
+At the most fundamental level, a coin selection algorithm is a _mathematical
+function_ that when applied to a [requested output
+set](#requested-output-set) and an [initial UTxO set](#initial-utxo-set),
+will produce a [coin selection](#coin-selection): the basis for a
+[transaction](#transaction) in a UTxO-based blockchain.
+
+This section describes:
+
+  * the [parameters](#parameters) accepted by all coin selection algorithms;
+  * the [results](#results) they produce when successful;
+  * the [error conditions](#failure-modes) that may occur on failure;
+  * the [properties](#properties) that apply to all coin selection
+    algorithms: mathematical laws governing the relationships between parameters
+    and results.
+
+In this section, the terms _coin selection algorithm_ and _coin selection
+function_ will be used interchangeably.
+
+## Parameters
+
+All coin selection functions accept the following parameters:
+
+ 1. #### Requested Output Set
+
+    A list of payments to be made to recipient addresses, encoded as a list of
+    [transaction outputs](#transaction-output).
+
+ 2. #### Initial UTxO Set
+
+    A [UTxO set](#utxo-set) from which the coin selection algorithm can select
+    entries, to cover payments listed in the [requested output
+    set](#requested-output-set).
+
+    In the context of a wallet, this parameter would normally be assigned with
+    the wallet's complete [UTxO set](#utxo-set), giving the coin selection
+    algorithm access to the total value associated with that wallet.
+
+ 3. #### Maximum Input Count
+
+    An _upper bound_ on the number of UTxO entries that the coin selection
+    algorithm is permitted to select from the [initial UTxO
+    set](#initial-utxo-set).
+
+    This parameter is necessary for blockchains that impose an upper limit on
+    the size of transactions.
+
+## Results
+
+All coin selection functions produce the following result values:
+
+ 1. #### Coin Selection
+
+    A _coin selection_ is the basis for a [transaction](#transaction) in a
+    UTxO-based blockchain.
+
+    It is a record with three fields:
+
+      * A set of **_inputs_**, equivalent to a subset of the
+        [initial UTxO set](#initial-utxo-set).
+
+        From the point of view of a _wallet_, this represents the value that
+        has been selected from the wallet in order to cover the total payment
+        value.
+
+      * A set of **_outputs_** (see [transaction output](#transaction-output)).
+
+        Represents the set of payments to be made to recipient addresses.
+
+      * A set of **_change values_** (see [change output](#change-output)),
```

Comment:
Do we have a minimum size for the value of a change? For example, if the change is just one lovelace, maybe it should just be added to the fee instead.

Reply:
> Do we have a minimum size for the value of a change? For example, if the change is just one lovelace, maybe it should just be added to the fee instead.

Yes, we do. We have the concept of a [dust threshold](https://input-output-hk.github.io/cardano-coin-selection/haddock/cardano-coin-selection-1.0.1/Cardano-CoinSelection-Fee.html#t:DustThreshold).

Here's an extract from the docs:
> A `DustThreshold` defines the _maximum_ size of a dust coin.
>
> Functions that accept a `DustThreshold` argument will generally exclude values that are less than or equal to this threshold from the change sets of generated selections, coalescing such coins together into larger coins that _exceed_ the threshold.

This dust threshold parameter is used in the _fee adjustment phase_ (see https://github.com/cardano-foundation/CIPs/pull/2#discussion_r464806172), which not only adjusts a coin selection in order to pay for the network fee, but also handles the process of coalescing dust coins into larger coins that _exceed_ the dust threshold.